File: crates/captain-api/static/js/app/views/NativeCapabilities.js
import { h } from 'preact';
import { useCallback, useEffect, useState } from 'preact/hooks';
import htm from 'htm';
import { api } from '../api.js';
import { toast } from '../store.js';

const html = htm.bind(h);
const STARTER_SOURCE = `format = 1
name = "project-summary"
description = "Read the project README and manifest in parallel."
version = "1.0.0"
output = { readme = "{{steps.readme.output}}", manifest = "{{steps.manifest.output}}" }

[inputs.root]
type = "string"
description = "Project root"

[permissions]
tools = ["file_read"]
read_paths = ["{{input.root}}/**"]

[policy]
timeout_secs = 60
max_parallel = 2

[[steps]]
id = "readme"
tool = "file_read"
needs = []
with = { path = "{{input.root}}/README.md" }

[[steps]]
id = "manifest"
tool = "file_read"
needs = []
with = { path = "{{input.root}}/Cargo.toml" }
`;

export function NativeCapabilitiesTab() {
  const [scope, setScope] = useState('effective');
  const [workspace, setWorkspace] = useState('');
  const [query, setQuery] = useState('');
  const [data, setData] = useState(null);
  const [runs, setRuns] = useState(null);
  const [selected, setSelected] = useState(null);
  const [showForge, setShowForge] = useState(false);

  const load = useCallback(async () => {
    if (scope === 'project' && !workspace.trim()) {
      setData({ count: 0, capabilities: [] });
      return;
    }
    try {
      const params = { scope, workspace: workspace.trim() || undefined };
      const [capabilities, recentRuns] = await Promise.all([
        api.nativeCapabilities(params),
        api.nativeCapabilityRuns(20),
      ]);
      setData(capabilities);
      setRuns(recentRuns);
    } catch (error) {
      toast(`Chargement impossible : ${error.message}`, 'err');
    }
  }, [scope, workspace]);

  useEffect(() => { load(); }, [load]);

  const capabilities = sortCapabilities((data && data.capabilities) || []);
  const needle = query.trim().toLocaleLowerCase('fr');
  const filtered = needle
    ? capabilities.filter((item) => `${item.name} ${item.description || ''} ${item.status}`
      .toLocaleLowerCase('fr').includes(needle))
    : capabilities;
  const pending = capabilities.filter((item) => item.human_action_required).length;
  const operational = capabilities.filter((item) => item.ready).length;

  return html`
    <div class="native-capabilities">
      <div class="capability-toolbar native-capability-toolbar">
        <div class="metrics-row">
          <div class="metric-chip">${data ? data.count : '…'} native(s)</div>
          <div class="metric-chip ${operational ? 'ok' : ''}">${operational} prête(s)</div>
          <div class="metric-chip ${pending ? 'off' : ''}">${pending} décision(s)</div>
        </div>
        <div class="native-toolbar-actions">
          <button class="ghost" onClick=${load}>Actualiser</button>
          <button onClick=${() => setShowForge((visible) => !visible)}>
            ${showForge ? 'Fermer' : 'Installer'}
          </button>
        </div>
      </div>

      <div class="native-scope-bar">
        <select value=${scope} onChange=${(event) => setScope(event.target.value)} aria-label="Portée">
          <option value="effective">Effectif</option>
          <option value="global">Global</option>
          <option value="project">Projet</option>
          <option value="all">Global + projet</option>
        </select>
        <input type="text" value=${workspace} placeholder="Workspace projet"
          onInput=${(event) => setWorkspace(event.target.value)} />
        <input type="text" value=${query} placeholder="Filtrer les capacités"
          onInput=${(event) => setQuery(event.target.value)} />
      </div>

      ${showForge && html`<${ForgePanel} workspace=${workspace} onInstalled=${async () => {
        setShowForge(false);
        await load();
      }} />`}

      ${data === null && html`<div class="skeleton" style="height:90px"></div>`}
      ${data && scope === 'project' && !workspace.trim() && html`
        <div class="status-empty">Renseignez le workspace du projet.</div>
      `}
      ${data && filtered.length === 0 && !(scope === 'project' && !workspace.trim()) && html`
        <div class="empty-state"><div class="glyph">◇</div><div>Aucune capacité native dans cette portée.</div></div>
      `}
      ${filtered.length > 0 && html`
        <div class="item-list native-capability-list">
          ${filtered.map((item) => html`
            <${CapabilityRow}
              key=${`${item.scope}:${item.name}:${item.selected_hash || ''}`}
              item=${item}
              workspace=${workspace}
              open=${selected && selected.name === item.name && selected.scope === item.scope}
              onToggle=${() => setSelected((current) => current && current.name === item.name
                && current.scope === item.scope ? null : item)}
              onChanged=${load}
            />
          `)}
        </div>
      `}

      <${RunsPanel} data=${runs} onChanged=${load} />
    </div>
  `;
}

function ForgePanel({ workspace, onInstalled }) {
  const [source, setSource] = useState(STARTER_SOURCE);
  const [name, setName] = useState('');
  const [scope, setScope] = useState('global');
  const [validation, setValidation] = useState(null);
  const [busy, setBusy] = useState(false);

  const validate = async () => {
    setBusy(true);
    try {
      const result = await api.validateNativeCapability({
        source,
        name: name.trim() || null,
      });
      setValidation(result);
      toast(`CapSpec ${shortHash(result.source_hash)} valide.`, 'ok');
    } catch (error) {
      setValidation(null);
      toast(`CapSpec invalide : ${error.message}`, 'err');
    } finally { setBusy(false); }
  };

  const install = async () => {
    if (scope === 'project' && !workspace.trim()) {
      toast('Le workspace projet est requis.', 'err');
      return;
    }
    setBusy(true);
    try {
      const result = await api.installNativeCapability({
        source,
        name: name.trim() || null,
        scope,
        workspace: scope === 'project' ? workspace.trim() : null,
      });
      toast(result.human_action_required
        ? `Proposition ${result.name} enregistrée : approbation requise.`
        : `Capacité ${result.name} prête.`, 'ok');
      await onInstalled();
    } catch (error) {
      toast(`Installation refusée : ${error.message}`, 'err');
    } finally { setBusy(false); }
  };

  return html`
    <section class="native-forge" aria-label="Installer une capacité native">
      <div class="native-forge-head">
        <h2>Fichier .captain</h2>
        ${validation && html`
          <span class="status-pill status-done">valide · ${shortHash(validation.source_hash)}</span>
        `}
      </div>
      <div class="native-forge-controls">
        <input type="text" value=${name} placeholder="Nom attendu (optionnel)"
          onInput=${(event) => { setName(event.target.value); setValidation(null); }} />
        <select value=${scope} onChange=${(event) => setScope(event.target.value)}>
          <option value="global">Global</option>
          <option value="project">Projet courant</option>
        </select>
      </div>
      <textarea class="native-source-editor" spellcheck="false" value=${source}
        onInput=${(event) => { setSource(event.target.value); setValidation(null); }} />
      <div class="form-actions">
        <button class="ghost" disabled=${busy} onClick=${validate}>Valider</button>
        <button disabled=${busy} onClick=${install}>Installer</button>
      </div>
    </section>
  `;
}

function CapabilityRow({ item, workspace, open, onToggle, onChanged }) {
  return html`
    <div class="native-capability-entry">
      <div class="item-row">
        <button class="native-row-toggle" onClick=${onToggle} aria-expanded=${open}>
          <span class="native-row-title">
            <strong>${item.name}</strong>
            <span class="item-row-meta">${item.description || item.tool_name}</span>
            <span class="item-row-meta">
              ${item.version ? `v${item.version} · ` : ''}${item.scope} · ${shortHash(item.selected_hash)}
            </span>
          </span>
          <span class="status-pill ${statusTone(item.status)}">${statusLabel(item.status)}</span>
        </button>
        ${item.human_action_required && html`
          <${DecisionButtons} item=${item} workspace=${workspace} onChanged=${onChanged} />
        `}
      </div>
      ${open && html`<${CapabilityDetail} item=${item} workspace=${workspace} onChanged=${onChanged} />`}
    </div>
  `;
}

function DecisionButtons({ item, workspace, onChanged }) {
  const [busy, setBusy] = useState(false);
  const decide = async (decision) => {
    setBusy(true);
    try {
      await api.decideNativeCapability(item.name, {
        decision,
        expected_hash: item.pending_hash,
        scope: item.scope,
        workspace: item.scope === 'project' ? workspace.trim() : null,
      });
      toast(decision === 'approve' ? `${item.name} approuvée.` : `${item.name} refusée.`, 'ok');
      await onChanged();
    } catch (error) {
      toast(`Décision impossible : ${error.message}`, 'err');
    } finally { setBusy(false); }
  };
  return html`
    <div class="item-row-actions">
      <button class="ghost danger" disabled=${busy} onClick=${() => decide('reject')}>Refuser</button>
      <button disabled=${busy} onClick=${() => decide('approve')}>Approuver le hash</button>
    </div>
  `;
}

function CapabilityDetail({ item, workspace, onChanged }) {
  const [detail, setDetail] = useState(null);
  const [sourceVisible, setSourceVisible] = useState(false);
  const [busy, setBusy] = useState(false);
  const [disableArmed, setDisableArmed] = useState(false);

  const load = useCallback(async (includeSource = false) => {
    try {
      const result = await api.nativeCapability(item.name, {
        scope: item.scope,
        workspace: item.scope === 'project' ? workspace.trim() : undefined,
        include_source: includeSource,
      });
      setDetail(result);
      setSourceVisible(includeSource);
    } catch (error) {
      toast(`Inspection impossible : ${error.message}`, 'err');
    }
  }, [item.name, item.scope, workspace]);

  useEffect(() => { load(false); }, [load]);

  const rollback = async (hash) => {
    setBusy(true);
    try {
      await api.rollbackNativeCapability(item.name, {
        target_hash: hash,
        scope: item.scope,
        workspace: item.scope === 'project' ? workspace.trim() : null,
      });
      toast(`${item.name} restaurée sur ${shortHash(hash)}.`, 'ok');
      await load(false);
      await onChanged();
    } catch (error) {
      toast(`Rollback impossible : ${error.message}`, 'err');
    } finally { setBusy(false); }
  };

  const disable = async () => {
    if (!disableArmed) {
      setDisableArmed(true);
      return;
    }
    setBusy(true);
    try {
      await api.disableNativeCapability(item.name, {
        scope: item.scope,
        workspace: item.scope === 'project' ? workspace.trim() : undefined,
      });
      toast(`${item.name} désactivée ; historique conservé.`, 'ok');
      await onChanged();
    } catch (error) {
      toast(`Désactivation impossible : ${error.message}`, 'err');
    } finally { setBusy(false); setDisableArmed(false); }
  };

  if (!detail) return html`<div class="detail-panel"><div class="skeleton" style="height:70px"></div></div>`;
  return html`
    <div class="native-capability-detail">
      <div class="native-detail-grid">
        <div><span>Tool</span><strong>${detail.tool_name}</strong></div>
        <div><span>Hash actif</span><strong>${shortHash(detail.active_hash)}</strong></div>
        <div><span>Empreinte permissions</span><strong>${shortHash(detail.permission_fingerprint)}</strong></div>
        <div><span>Étapes</span><strong>${(detail.steps || []).length}</strong></div>
      </div>
      <div class="native-permissions">
        ${(detail.permissions && detail.permissions.tools || []).map((tool) => html`
          <span class="status-pill status-review" key=${tool}>${tool}</span>
        `)}
      </div>
      ${detail.last_error && html`<div class="status-issue tone-err"><div><strong>Erreur source</strong><span>${detail.last_error}</span></div></div>`}
      <div class="native-detail-actions">
        <button class="ghost" disabled=${busy} onClick=${() => load(!sourceVisible)}>
          ${sourceVisible ? 'Masquer la source' : 'Voir la source'}
        </button>
        <button class="ghost danger" disabled=${busy} onClick=${disable}>
          ${disableArmed ? 'Confirmer la désactivation' : 'Désactiver'}
        </button>
      </div>
      ${sourceVisible && html`<pre class="code-block">${detail.source || ''}</pre>`}
      <div class="native-revisions">
        <h3>Révisions</h3>
        ${(detail.revisions || []).map((revision) => html`
          <div class="native-revision" key=${revision.source_hash}>
            <div>
              <strong>${revision.version || 'sans version'} · ${shortHash(revision.source_hash)}</strong>
              <span>${revision.approved_by ? `approuvée par ${revision.approved_by}`
                : revision.rejected_by ? `refusée par ${revision.rejected_by}` : 'non décidée'}</span>
            </div>
            ${revision.source_hash !== detail.active_hash && html`
              <button class="ghost" disabled=${busy} onClick=${() => rollback(revision.source_hash)}>Restaurer</button>
            `}
          </div>
        `)}
      </div>
    </div>
  `;
}

function RunsPanel({ data, onChanged }) {
  const runs = (data && data.runs) || [];
  return html`
    <section class="native-runs">
      <h2>Runs récents</h2>
      ${data === null && html`<div class="skeleton" style="height:54px"></div>`}
      ${data && runs.length === 0 && html`<div class="status-empty">Aucun run natif.</div>`}
      ${runs.map((run) => html`<${RunRow} key=${run.run_id} run=${run} onChanged=${onChanged} />`)}
    </section>
  `;
}

function RunRow({ run, onChanged }) {
  const [open, setOpen] = useState(false);
  const uncertain = (run.nodes || []).find((node) => node.status === 'uncertain');
  return html`
    <div class="native-run-entry">
      <div class="native-run">
        <div>
          <strong>${run.capability_name}</strong>
          <span>${shortHash(run.run_id)} · ${shortHash(run.source_hash)} · ${run.origin || 'runtime'}</span>
        </div>
        <div class="native-run-actions">
          <span class="status-pill ${runTone(run.status)}">${String(run.status).replaceAll('_', ' ')}</span>
          ${uncertain && html`
            <button class="ghost" onClick=${() => setOpen((value) => !value)} aria-expanded=${open}>
              ${open ? 'Fermer' : 'Décider'}
            </button>
          `}
        </div>
      </div>
      ${open && uncertain && html`
        <${RunDecisionPanel} run=${run} node=${uncertain} onChanged=${async () => {
          setOpen(false);
          await onChanged();
        }} />
      `}
    </div>
  `;
}

function RunDecisionPanel({ run, node, onChanged }) {
  const [output, setOutput] = useState('null');
  const [reason, setReason] = useState('');
  const [busy, setBusy] = useState(false);

  const decide = async (decision) => {
    let parsedOutput;
    if (decision === 'confirm_succeeded') {
      try { parsedOutput = JSON.parse(output); } catch {
        toast('La sortie confirmée doit être du JSON valide.', 'err');
        return;
      }
    }
    if (decision === 'mark_failed' && !reason.trim()) {
      toast('Le motif d’échec est requis.', 'err');
      return;
    }
    setBusy(true);
    try {
      const body = {
        node_id: node.step_id,
        expected_tool_use_id: node.tool_use_id,
        expected_attempt: node.attempts,
        decision,
      };
      if (decision === 'confirm_succeeded') body.output = parsedOutput;
      if (decision === 'mark_failed') body.reason = reason.trim();
      const result = await api.resolveNativeCapabilityRun(run.run_id, body);
      toast(result.resume_scheduled
        ? `Décision acceptée pour ${node.step_id} ; reprise planifiée.`
        : `Run ${shortHash(run.run_id)} marqué en échec.`, 'ok');
      await onChanged();
    } catch (error) {
      toast(`Décision refusée : ${error.message}`, 'err');
      await onChanged();
    } finally { setBusy(false); }
  };

  return html`
    <div class="native-run-decision">
      <div class="native-detail-grid">
        <div><span>Étape</span><strong>${node.step_id}</strong></div>
        <div><span>Tool</span><strong>${node.tool_name}</strong></div>
        <div><span>Tentative</span><strong>${node.attempts}</strong></div>
        <div><span>Tool use</span><strong>${shortHash(node.tool_use_id)}</strong></div>
      </div>
      <div class="native-run-decision-fields">
        <label>Sortie observée (JSON)
          <textarea spellcheck="false" value=${output}
            onInput=${(event) => setOutput(event.target.value)} />
        </label>
        <label>Motif d’échec
          <input type="text" value=${reason}
            onInput=${(event) => setReason(event.target.value)} />
        </label>
      </div>
      <div class="native-detail-actions">
        <button class="ghost danger" disabled=${busy} onClick=${() => decide('mark_failed')}>Marquer échoué</button>
        <button class="ghost" disabled=${busy} onClick=${() => decide('retry')}>Réessayer</button>
        <button disabled=${busy} onClick=${() => decide('confirm_succeeded')}>Confirmer réussi</button>
      </div>
    </div>
  `;
}

function sortCapabilities(items) {
  const priority = (item) => item.human_action_required ? 0 : item.ready ? 1 : 2;
  return items.slice().sort((left, right) => priority(left) - priority(right)
    || String(left.name).localeCompare(String(right.name)));
}

function shortHash(hash) {
  if (!hash) return '—';
  return String(hash).slice(0, 12);
}

function statusLabel(status) {
  return ({
    operational: 'prête',
    pending_approval: 'à approuver',
    update_pending_approval: 'mise à jour à approuver',
    invalid: 'invalide',
    invalid_update_retained: 'actif · mise à jour invalide',
    disabled: 'désactivée',
    rejected: 'refusée',
    update_rejected: 'mise à jour refusée',
  })[status] || status;
}

function statusTone(status) {
  if (status === 'operational' || status === 'invalid_update_retained') return 'status-done';
  if (status === 'pending_approval' || status === 'update_pending_approval') return 'status-review';
  return 'status-cancelled';
}

function runTone(status) {
  if (status === 'succeeded') return 'status-done';
  if (status === 'running' || status === 'waiting_decision') return 'status-review';
  if (status === 'failed') return 'status-blocked';
  return 'status-paused';
}

File: crates/captain-cli/src/tui/screens/native_capabilities.rs
//! Native CapSpec operator screen: exact-hash decisions and durable revisions.

use ratatui::crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use ratatui::widgets::ListState;

mod render;
pub use render::draw;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum NativeScope {
    Effective,
    Global,
    Project,
}

impl NativeScope {
    pub fn query(self) -> &'static str {
        match self {
            Self::Effective => "effective",
            Self::Global => "global",
            Self::Project => "project",
        }
    }

    fn label(self) -> &'static str {
        match self {
            Self::Effective => "effective",
            Self::Global => "global",
            Self::Project => "project",
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct NativeRevision {
    pub source_hash: String,
    pub version: String,
    pub approved_by: Option<String>,
    pub rejected_by: Option<String>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct NativeCapabilityInfo {
    pub name: String,
    pub tool_name: String,
    pub description: String,
    pub version: String,
    pub status: String,
    pub scope: String,
    pub ready: bool,
    pub human_action_required: bool,
    pub active_hash: Option<String>,
    pub pending_hash: Option<String>,
    pub selected_hash: Option<String>,
    pub permission_fingerprint: String,
    pub tools: Vec<String>,
    pub revisions: Vec<NativeRevision>,
    pub source: Option<String>,
    pub last_error: Option<String>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct NativeRunInfo {
    pub run_id: String,
    pub capability_name: String,
    pub source_hash: String,
    pub status: String,
    pub origin: String,
    pub nodes: Vec<NativeRunNode>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct NativeRunNode {
    pub step_id: String,
    pub tool_name: String,
    pub status: String,
    pub attempts: u32,
    pub tool_use_id: Option<String>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum NativeRunDecision {
    ConfirmSucceeded,
    Retry,
    MarkFailed,
}

pub struct NativeCapabilitiesState {
    pub scope: NativeScope,
    pub capabilities: Vec<NativeCapabilityInfo>,
    pub runs: Vec<NativeRunInfo>,
    pub list_state: ListState,
    pub revision_index: usize,
    pub run_index: usize,
    pub loading: bool,
    pub source_visible: bool,
    pub confirm_disable: bool,
    pub confirm_run: Option<NativeRunDecision>,
    pub status_msg: String,
}

#[derive(Debug, PartialEq, Eq)]
pub enum NativeCapabilitiesAction {
    Continue,
    Refresh,
    Inspect {
        name: String,
        scope: String,
        include_source: bool,
    },
    Decide {
        name: String,
        scope: String,
        expected_hash: String,
        approve: bool,
    },
    Rollback {
        name: String,
        scope: String,
        target_hash: String,
    },
    Disable {
        name: String,
        scope: String,
    },
    ResolveRun {
        run_id: String,
        node_id: String,
        tool_use_id: String,
        attempt: u32,
        decision: NativeRunDecision,
    },
}

impl NativeCapabilitiesState {
    pub fn new() -> Self {
        Self {
            scope: NativeScope::Effective,
            capabilities: Vec::new(),
            runs: Vec::new(),
            list_state: ListState::default(),
            revision_index: 0,
            run_index: 0,
            loading: false,
            source_visible: false,
            confirm_disable: false,
            confirm_run: None,
            status_msg: String::new(),
        }
    }

    pub fn replace(
        &mut self,
        mut capabilities: Vec<NativeCapabilityInfo>,
        runs: Vec<NativeRunInfo>,
    ) {
        let selected_name = self.selected().map(|item| item.name.clone());
        capabilities.sort_by(|left, right| {
            capability_priority(left)
                .cmp(&capability_priority(right))
                .then_with(|| left.name.cmp(&right.name))
        });
        self.capabilities = capabilities;
        self.runs = runs;
        let waiting = self.uncertain_run_count();
        self.run_index = self.run_index.min(waiting.saturating_sub(1));
        let selected = selected_name
            .and_then(|name| self.capabilities.iter().position(|item| item.name == name))
            .or((!self.capabilities.is_empty()).then_some(0));
        self.list_state.select(selected);
        self.revision_index = 0;
        self.loading = false;
    }

    pub fn replace_inspected(&mut self, inspected: NativeCapabilityInfo) {
        if let Some(index) = self
            .capabilities
            .iter()
            .position(|item| item.name == inspected.name && item.scope == inspected.scope)
        {
            self.capabilities[index] = inspected;
            self.list_state.select(Some(index));
        }
        self.loading = false;
    }

    pub fn selected(&self) -> Option<&NativeCapabilityInfo> {
        self.list_state
            .selected()
            .and_then(|index| self.capabilities.get(index))
    }

    pub fn handle_key(&mut self, key: KeyEvent) -> NativeCapabilitiesAction {
        if key.code == KeyCode::Char('c') && key.modifiers.contains(KeyModifiers::CONTROL) {
            return NativeCapabilitiesAction::Continue;
        }
        if self.confirm_disable {
            return match key.code {
                KeyCode::Char('y') | KeyCode::Char('Y') => {
                    self.confirm_disable = false;
                    self.selected_action(|item| NativeCapabilitiesAction::Disable {
                        name: item.name.clone(),
                        scope: item.scope.clone(),
                    })
                }
                _ => {
                    self.confirm_disable = false;
                    NativeCapabilitiesAction::Continue
                }
            };
        }
        if let Some(decision) = self.confirm_run.take() {
            return match key.code {
                KeyCode::Char('y') | KeyCode::Char('Y') => self.run_decision_action(decision),
                _ => NativeCapabilitiesAction::Continue,
            };
        }
        match key.code {
            KeyCode::Char('1') => return self.switch_scope(NativeScope::Effective),
            KeyCode::Char('2') => return self.switch_scope(NativeScope::Global),
            KeyCode::Char('3') => return self.switch_scope(NativeScope::Project),
            KeyCode::Char('r') => return NativeCapabilitiesAction::Refresh,
            KeyCode::Up | KeyCode::Char('k') => self.move_selection(false),
            KeyCode::Down | KeyCode::Char('j') => self.move_selection(true),
            KeyCode::Left | KeyCode::Char('h') => self.move_revision(false),
            KeyCode::Right | KeyCode::Char('l') => self.move_revision(true),
            KeyCode::Char('[') => self.move_run(false),
            KeyCode::Char(']') => self.move_run(true),
            KeyCode::Enter => {
                return self.selected_action(|item| NativeCapabilitiesAction::Inspect {
                    name: item.name.clone(),
                    scope: item.scope.clone(),
                    include_source: false,
                });
            }
            KeyCode::Char('v') => {
                let include_source = !self.source_visible;
                self.source_visible = include_source;
                if include_source
                    && self
                        .selected()
                        .and_then(|item| item.source.as_ref())
                        .is_none()
                {
                    return self.selected_action(|item| NativeCapabilitiesAction::Inspect {
                        name: item.name.clone(),
                        scope: item.scope.clone(),
                        include_source: true,
                    });
                }
            }
            KeyCode::Char('a') => return self.decision_action(true),
            KeyCode::Char('x') => return self.decision_action(false),
            KeyCode::Char('b') => return self.rollback_action(),
            KeyCode::Char('d') | KeyCode::Delete => {
                if self.selected().is_some() {
                    self.confirm_disable = true;
                }
            }
            KeyCode::Char('T') => self.arm_run_decision(NativeRunDecision::Retry),
            KeyCode::Char('C') => self.arm_run_decision(NativeRunDecision::ConfirmSucceeded),
            KeyCode::Char('F') => self.arm_run_decision(NativeRunDecision::MarkFailed),
            _ => {}
        }
        NativeCapabilitiesAction::Continue
    }

    fn switch_scope(&mut self, scope: NativeScope) -> NativeCapabilitiesAction {
        self.scope = scope;
        self.loading = true;
        self.status_msg.clear();
        NativeCapabilitiesAction::Refresh
    }

    fn move_selection(&mut self, forward: bool) {
        if self.capabilities.is_empty() {
            return;
        }
        let current = self.list_state.selected().unwrap_or(0);
        let next = if forward {
            (current + 1) % self.capabilities.len()
        } else if current == 0 {
            self.capabilities.len() - 1
        } else {
            current - 1
        };
        self.list_state.select(Some(next));
        self.revision_index = 0;
        self.source_visible = false;
        self.confirm_disable = false;
        self.confirm_run = None;
    }

    fn move_run(&mut self, forward: bool) {
        let total = self.uncertain_run_count();
        if total == 0 {
            return;
        }
        self.run_index = if forward {
            (self.run_index + 1) % total
        } else if self.run_index == 0 {
            total - 1
        } else {
            self.run_index - 1
        };
        self.confirm_run = None;
    }

    fn arm_run_decision(&mut self, decision: NativeRunDecision) {
        if self.selected_uncertain().is_some() {
            self.confirm_run = Some(decision);
        }
    }

    pub fn selected_uncertain(&self) -> Option<(&NativeRunInfo, &NativeRunNode)> {
        self.runs
            .iter()
            .filter_map(|run| {
                (run.status == "waiting_decision")
                    .then(|| {
                        run.nodes
                            .iter()
                            .find(|node| node.status == "uncertain")
                            .map(|node| (run, node))
                    })
                    .flatten()
            })
            .nth(self.run_index)
    }

    pub fn uncertain_run_count(&self) -> usize {
        self.runs
            .iter()
            .filter(|run| {
                run.status == "waiting_decision"
                    && run.nodes.iter().any(|node| node.status == "uncertain")
            })
            .count()
    }

    fn run_decision_action(&self, decision: NativeRunDecision) -> NativeCapabilitiesAction {
        self.selected_uncertain()
            .and_then(|(run, node)| {
                node.tool_use_id
                    .as_ref()
                    .map(|tool_use_id| NativeCapabilitiesAction::ResolveRun {
                        run_id: run.run_id.clone(),
                        node_id: node.step_id.clone(),
                        tool_use_id: tool_use_id.clone(),
                        attempt: node.attempts,
                        decision,
                    })
            })
            .unwrap_or(NativeCapabilitiesAction::Continue)
    }

    fn move_revision(&mut self, forward: bool) {
        let total = self
            .selected()
            .map(|item| item.revisions.len())
            .unwrap_or(0);
        if total == 0 {
            return;
        }
        self.revision_index = if forward {
            (self.revision_index + 1) % total
        } else if self.revision_index == 0 {
            total - 1
        } else {
            self.revision_index - 1
        };
    }

    fn decision_action(&self, approve: bool) -> NativeCapabilitiesAction {
        self.selected_action(|item| match item.pending_hash.as_ref() {
            Some(hash) if item.human_action_required => NativeCapabilitiesAction::Decide {
                name: item.name.clone(),
                scope: item.scope.clone(),
                expected_hash: hash.clone(),
                approve,
            },
            _ => NativeCapabilitiesAction::Continue,
        })
    }

    fn rollback_action(&self) -> NativeCapabilitiesAction {
        self.selected_action(|item| {
            item.revisions
                .get(self.revision_index)
                .filter(|revision| {
                    revision.approved_by.is_some()
                        && Some(&revision.source_hash) != item.active_hash.as_ref()
                })
                .map(|revision| NativeCapabilitiesAction::Rollback {
                    name: item.name.clone(),
                    scope: item.scope.clone(),
                    target_hash: revision.source_hash.clone(),
                })
                .unwrap_or(NativeCapabilitiesAction::Continue)
        })
    }

    fn selected_action(
        &self,
        action: impl FnOnce(&NativeCapabilityInfo) -> NativeCapabilitiesAction,
    ) -> NativeCapabilitiesAction {
        self.selected()
            .map(action)
            .unwrap_or(NativeCapabilitiesAction::Continue)
    }
}

fn capability_priority(item: &NativeCapabilityInfo) -> u8 {
    if item.human_action_required {
        0
    } else if item.ready {
        1
    } else {
        2
    }
}

#[cfg(test)]
#[path = "native_capabilities/tests.rs"]
mod tests;
